## Preprocessing Single Cell RNA-seq Data in Seurat

This tutorial guides you through preprocessing steps for single-cell RNA-seq data using Seurat. We'll cover:

* Loading data from Cell Ranger output
* Creating a Seurat object with basic filtering
* Removing ambient RNA with SoupX
* Identifying and removing doublets with scDblFinder
* Filtering low-quality and apoptotic cells

**packages:**

* Seurat
* SoupX
* scDblFinder

**Data:**

After running cellranger you can navigate to the outs folder, in the outs folder you will find 'filtered_feature_bc_matrix' which contains the filtered gene expression matrix.

Once you start Rstudio you can navigate to the filtered_feature_bc_matrix folder using the 'Files' tab in Rstudio 

Once in the correct directory, you can copy it using the More menu,

Now we have to load the libraries we are going to use in our analysis
```R
# Load libraries
library(dplyr)
library(Seurat)
library(SoupX)
```

# 1. Load data from cellranger
```R
pbmc.data <- Read10X(data.dir = "/project/ag-singlecellsummer2024/ahmedhas_cellrangeroutputs/outs/filtered_feature_bc_matrix/")
```

# 2. Create Seurat object 
```R
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "10k", min.cells = 3, min.features = 200)
```

# 3. Remove ambient RNA with SoupX
```R
sc <- load10X("/project/ag-singlecellsummer2024/ahmedhas_cellrangeroutputs/outs/")
sc <- autoEstCont(sc)
out <- adjustCounts(sc)
pbmc0 <- CreateSeuratObject(counts = out, project = "10k", min.cells = 3, min.features = 200)
```

**Check SoupX removal efficiency**
```R
removed_counts <- sum(pbmc$nCount_RNA) - sum(pbmc0$nCount_RNA)
cat("SoupX removed", removed_counts, "gene counts as ambient RNA.\n")
```

# 4. Identify and remove doublets with scDblFinder
```R
scd <- as.SingleCellExperiment(pbmc0)
scd <- scDblFinder(scd)
```

**Analyze doublet scores**
```R
results <- data.frame(
  "Barcode" = rownames(colData(scd)),
  "scDblFinder_DropletType" = scd$scDblFinder.class,
  "scDblFinder_Score" = scd$scDblFinder.score)
summary <- table(results$scDblFinder_DropletType)
per <- prop.table(summary) * 100
pie(a, col = c("whitesmoke", "indianred1"), main = "Percentage doublets", labels = paste(rownames(a)," ", round(a, 2), "% ", "(", summary$`Droplet Number`, " cells)",  sep = ""))
```

**Subset Seurat object to singlets**
```R
pbmc0$Droplet_type <- results$scDblFinder_DropletType
pbmc0 <- subset(x = pbmc0, subset = Droplet_type == "singlet")
```

# 5. Removing low-quality and apoptotic cells
```R
pbmc0[["percent.mt"]] <- PercentageFeatureSet(pbmc0, pattern = "^MT-")
pbmc0[["percent.ribo"]] <- Seurat::PercentageFeatureSet(pbmc0, pattern = "^RP[SL]")
pbmc0[["percent.globin"]] <- Seurat::PercentageFeatureSet(pbmc0, pattern = "^HB[^(P)]")
```

**Visualize QC metrics**
```R
VlnPlot(pbmc0, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.ribo". "percent.globin" ), ncol = 5)
```

**Explore relationships between metrics with scatter plots**
```R
plot1 <- FeatureScatter(pbmc0, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc0, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

**Subset good cells and visualize QC metrics again**
```R
pbmc0 <- subset(pbmc0, subset = nFeature_RNA > 1000 & nFeature_RNA < 6000 & percent.mt < 10 & nCount_RNA < 30000)
VlnPlot(pbmc0, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.ribo". "percent.globin" ), ncol = 5)
```
This concludes our QC now the Seurat object is ready for the next steps



