## Preprocessing Single Cell RNA-seq Data in Seurat

This tutorial guides you through the preprocessing steps for single-cell RNA-seq data using Seurat. We'll cover:

* Loading data from Cell Ranger output
* Creating a Seurat object with basic filtering
* Removing ambient RNA with SoupX
* Identifying and removing doublets with scDblFinder
* Filtering low-quality and apoptotic cells

**packages:**

* [Seurat](https://satijalab.org/seurat/articles/pbmc3k_tutorial)
* [SoupX](https://github.com/constantAmateur/SoupX)
* [scDblFinder](https://github.com/plger/scDblFinder?tab=readme-ov-file)

**Data:**

After running cellranger you can navigate to the outs folder, in the outs folder you will find 'filtered_feature_bc_matrix' which contains the filtered gene expression matrix.

Once you start Rstudio you can navigate to the filtered_feature_bc_matrix folder using the 'Files' tab in Rstudio 

Once in the correct directory, you can copy it using the More menu,

Now we have to load the libraries we are going to use in our analysis
```R
# Load libraries
library(dplyr)
library(Seurat)
library(SoupX)
```

# 1. Load data from cellranger
```R
pbmc.data <- Read10X(data.dir = "/project/ag-singlecellsummer2024/ahmedhas_cellrangeroutputs/outs/filtered_feature_bc_matrix/")
```
If you look at loaded data you will get something like this

```R
pbmc.data
```

```bash
     
36601 x 11999 sparse Matrix of class "dgCMatrix"
  [[ suppressing 89 column names ‘AAACCCAAGGCCCAAA-1’, ‘AAACCCAAGTAATACG-1’, ‘AAACCCAAGTCACACT-1’ ... ]]
  [[ suppressing 89 column names ‘AAACCCAAGGCCCAAA-1’, ‘AAACCCAAGTAATACG-1’, ‘AAACCCAAGTCACACT-1’ ... ]]
                                                                                                                                                                                                    
MIR1302-2HG . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
FAM138A     . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
OR4F5       . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
AL627309.1  . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 2 . . . . 1 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
AL627309.3  . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
AL627309.2  . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......

 ..............................
 ........suppressing 11910 columns and 36590 rows in show(); maybe adjust options(max.print=, width=)
 ..............................
  [[ suppressing 89 column names ‘AAACCCAAGGCCCAAA-1’, ‘AAACCCAAGTAATACG-1’, ‘AAACCCAAGTCACACT-1’ ... ]]
                                                                                                                                                                                                   
AC141272.1 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
AC023491.2 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
AC007325.1 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
AC007325.4 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 1 . . . . . . . . . . . . . . . . . . . . . ......
AC007325.2 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
```
Here row names are gene names and col names are the individual cells, the matrix generated from the single-cell experiment is of a sparse type (meaning it contains lots of 0), We will use this matrix to create a Seurat object, and while at it we can also do a light QC
In the case you work with multiple datasets you can give a unique name to each one using 'project = "10k"'.

# 2. Create Seurat object 
Here I am only including genes that are at least expressed in three cells, and I am also including cells with at least 200 unique genes detected.
```R
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "10k", min.cells = 3, min.features = 200)
```
**Check cell/gene counts**

```R
pbmc@assays$RNA@layers$counts
```
You will get something like this, So we are down from 36601 genes x 11999 cells to 27363 genes x 11966 cells

```bash
27363 x 11966 sparse Matrix of class "dgCMatrix"
                                                                                                                                                                                                   
[1,] . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 2 . . . . 1 . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
[2,] . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
[3,] . . . . . . . . . . . 1 . . . . . . 1 . . . 1 . . . . . . . . . . . . . . . . . . . 1 . . . . 1 . . . . . . . . . . . . . . . . . . . . . 1 1 . . 1 . . . . . 1 . . . . . . . . . . . . ......
[4,] . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
[5,] . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......

 ..............................
 ........suppressing 11874 columns and 27353 rows in show(); maybe adjust options(max.print=, width=)
 ..............................
                                                                                                                                                                                                       
[27359,] 1 2 . 2 1 2 . 7 . . 4 1 . 2 . 1 2 . 1 3 . 2 . 3 1 3 2 . . . . . 1 . 1 1 . . 1 1 . . . 1 2 . . . . 1 . . 1 . 2 1 . . . 2 . . 1 . . 1 2 . . . . 1 . . 4 . 2 1 . . . . . . 1 . . 1 1 . . . ......
[27360,] . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 1 . . . . . . . . . . . . . . . . . . . . . . . . . . . 1 . . . . . . . . . . . . . . . . . . . . . . . . . . ......
[27361,] 1 1 . . 1 . 2 . . . . . . . . . . . . . . . . . . . . . . . . . 1 1 . . . . . . . . 1 . 1 . . . . . . . . . . 1 . . . . . 1 . . . . 1 . . . . . . . . . . 1 . . . . . . . . . . . . . . ......
[27362,] . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . ......
[27363,] . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . 1 . . . . . . . . . . . . . . . . . . . . . . . . ......

```

You can look at the pbmc Seurat object by running
```R
View(pbmc)
```


# 3. Remove ambient RNA with SoupX
The first QC step I do is usually getting rid of the ambient RNA to this we use 'SoupX' for more information follow this [link](https://github.com/constantAmateur/SoupX)


```R
sc <- load10X("/project/ag-singlecellsummer2024/ahmedhas_cellrangeroutputs/outs/")
sc <- autoEstCont(sc)
out <- adjustCounts(sc)
pbmc0 <- CreateSeuratObject(counts = out, project = "10k", min.cells = 3, min.features = 200)
```

**Check SoupX removal efficiency**
```R
removed_counts <- sum(pbmc$nCount_RNA) - sum(pbmc0$nCount_RNA)
cat("SoupX removed", removed_counts, "gene counts as ambient RNA.\n")
```
SoupX identified 1446930 gene counts as ambient RNA, now our new pbmc0 object should be ambient RNA free which I can use for the next step 

# 4. Identify and remove doublets with scDblFinder
There are multiple methods to get rid of doublet events more information [here](https://www.sciencedirect.com/science/article/pii/S2405471220304592?via%3Dihub)

I am here using 'scDblFinder' more information here [here](https://github.com/plger/scDblFinder?tab=readme-ov-file)
```R
scd <- as.SingleCellExperiment(pbmc0)
scd <- scDblFinder(scd)
```

**Analyze doublet scores**
```R
results <- data.frame("Barcode" = rownames(colData(scd)), "scDblFinder_DropletType" = scd$scDblFinder.class, "scDblFinder_Score" = scd$scDblFinder.score)
summary <- as.data.frame(table(results$scDblFinder_DropletType))
colnames(summary) <- c("Classification", "Droplet Number")
per <- prop.table(table(results$scDblFinder_DropletType))
a <- per*100
pie(a, col = c("whitesmoke", "indianred1"), main = "N4197 percentage doublets", labels = paste(rownames(a)," ", round(a, 2), "% ", "(", summary$`Droplet Number`, " cells)",  sep = ""))
```

**Subset Seurat object to singlets**
```R
pbmc0$Droplet_type <- results$scDblFinder_DropletType
pbmc0 <- subset(x = pbmc0, subset = Droplet_type == "singlet")
```

# 5. Removing low-quality and apoptotic cells
Most likely cells with high mitochondrial gene content are apoptotic so we need to remove them from our analysis.

We can calculate and visualise the QC metrics for several parameters such as  mitochondrial, ribosomal and haemoglobin genes.


```R
pbmc0[["percent.mt"]] <- PercentageFeatureSet(pbmc0, pattern = "^MT-")
pbmc0[["percent.ribo"]] <- Seurat::PercentageFeatureSet(pbmc0, pattern = "^RP[SL]")
pbmc0[["percent.globin"]] <- Seurat::PercentageFeatureSet(pbmc0, pattern = "^HB[^(P)]")
```

**Visualize QC metrics**

```R
VlnPlot(pbmc0, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.ribo", "percent.globin" ), ncol = 5)
```

**Explore relationships between metrics with scatter plots**
```R
plot1 <- FeatureScatter(pbmc0, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc0, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

**Subset good cells and visualize QC metrics again**
```R
pbmc0 <- subset(pbmc0, subset = nFeature_RNA > 1000 & nFeature_RNA < 6000 & percent.mt < 10 & nCount_RNA < 30000)
VlnPlot(pbmc0, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.ribo". "percent.globin" ), ncol = 5)
```
This concludes our QC now the Seurat object is ready for the next steps



